### ISSUE TITLE
Control flow operators issue tracker

### ISSUE THREAD

**Issue by ydwu4** [pytorch/pytorch#172568]
## Summary

This issue tracks unfinished work items for PyTorch's control flow higher-order operators (`cond`, `while_loop`, `scan`, `map`, `associative_scan`). If you're interested in working on any of them, please comment below!

## Background

PyTorch's control flow operators in `torch._higher_order_ops` enable expressing dynamic control flow in a way that's compatible with `torch.compile`, export, and other PyTorch 2.0 features. However, several features remain incomplete across these operators.

### 1. Input Mutation Support

**Context:** Input mutation allows the operator's subgraphs to mutate tensors passed as inputs, which is important for memory efficiency and compatibility with existing code patterns.
**Status:** Implemented for `invoke_subgraph`, needed for other operators

The following operators need input mutation support:
- [ ] `cond`
- [ ] `while_loop`
- [ ] `scan`

**Reference:**: https://github.com/pytorch/pytorch/pull/159009 and https://github.com/pytorch/pytorch/pull/154072/

---

### 2. DTensor Support

**Context:** DTensor is PyTorch's distributed tensor abstraction. Adding support enables these operators to work seamlessly in distributed training scenarios.
**Status:** Needs implementation for all control flow operators

The following operators need DTensor (Distributed Tensor) support:
- [ ] `cond`
- [ ] `while_loop`
- [ ] `scan`

**Reference:** https://github.com/pytorch/pytorch/pull/164548

---

### 3. Re-write vmap with `chunk_size > 1` Using Scan

**Status:** Carry draft pr to finish line.

**Context:** When `vmap` is called with `chunk_size > 1`, it should be expressible using `scan` for better performance and composability.

- [ ] Implement chunked vmap via scan

**Reference:**  https://github.com/pytorch/pytorch/pull/165616

---

### 4. Improve Scan Compilation Time
**Context:** There are a pretty high static overhead of compilation for scan.
**Status:** Needs investigation and optimization

This task involves:
- [ ] Profiling compilation to identify bottlenecks
- [ ] Implementing optimizations to reduce compile time

--- 

### 5. More efficient backward
**Context:** the backward now does a full re-compute of forward, which is sub-optimal.
**Status:** Needs investigation and optimization
This task involves:
- [ ] while_loop
- [ ] map
- [ ] associative_scan

**Reference:** https://github.com/pytorch/pytorch/pull/162754

---

### Versions

on master.

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @bdhirsh @bobrenjc93 @zou3519 @bohnstingl @galv @kshitij12345 @kiya00

**Comment by vadimkantorov** [comment]
Related on having a test/benchmark against other impls:
- https://github.com/pytorch/pytorch/pull/164859

**Comment by bohnstingl** [comment]
@ydwu4 I created the first PRs in a series of PRs to add the input mutations for:

- map: https://github.com/pytorch/pytorch/pull/183062
- scan: https://github.com/pytorch/pytorch/pull/183063
- associative_scan: https://github.com/pytorch/pytorch/pull/183064

**Comment by bohnstingl** [comment]
I also opened a PR to revive https://github.com/pytorch/pytorch/pull/165616, see https://github.com/pytorch/pytorch/pull/183182

**Comment by bohnstingl** [comment]
- https://github.com/pytorch/pytorch/pull/183182 was closed due to limited gains
- https://github.com/pytorch/pytorch/pull/183064 was closed because we currently do not want to support input mutation for `associative_scan`

Updated and ready for review:
- `map` input support: https://github.com/pytorch/pytorch/pull/183062
- `scan` input support: https://github.com/pytorch/pytorch/pull/183063
- `switch` first PR with skeleton: https://github.com/pytorch/pytorch/pull/182902

**Comment by yoyolicoris** [comment]
Hi, do the higher-order ops support `autograd`?
I asked this because I was trying to use `scan` for modal synthesis in a differentiable context, but I found that the gradients returned by `scan` were all zeros. Is that intentional, or did I miss something here?

Below is the minimum reproducible code. I'm using PyTorch 2.12.
cc @bohnstingl 

```python
import torch


def modal_ir(a1, a2, b1, n_samples):
    init_carry = (torch.zeros_like(b1), torch.zeros_like(b1))

    def step(carry, fin):
        q1, q2 = carry
        output = q1.sum()
        q_new = -a1 * q1 - a2 * q2 + b1 * fin
        return (q_new, q1.clone()), output

    xs = torch.zeros(n_samples)
    xs[0] = 1.0  # impulse at step 0
    _, ir = torch._higher_order_ops.scan(step, init_carry, xs)
    return ir


if __name__ == "__main__":

    a1 = torch.tensor([-0.5, -0.3], requires_grad=True)
    a2 = torch.tensor([0.29, 0.09], requires_grad=True)
    b1 = torch.tensor([1e-1, 1], requires_grad=True)  # avoid zero gain
    n_samples = 100
    ir = modal_ir(a1, a2, b1, n_samples)
    loss = ir.square().sum()
    loss.backward()
    print("a1 grad:", a1.grad)
    print("a2 grad:", a2.grad)
    print("b1 grad:", b1.grad)
```

**Comment by bohnstingl** [comment]
@yoyolicoris thank you for using `scan`!
The behavior that you describe is actually a known bug that was raised in this [issue](https://github.com/pytorch/pytorch/issues/184528) and for which a fix exists [here](https://github.com/pytorch/pytorch/pull/185332). However, it has not yet landed and in main, but hopefully will soon.

Until this is fixed, what you need to do is to ensure that your `init_c`arry has `requires_grad=True`. So doing
```
init_carry = (torch.zeros_like(b1, requires_grad=b1.requires_grad), 
              torch.zeros_like(b1, requires_grad=b1.requires_grad))
```
Should resolve your problem.
Please let me know and post here if you experience any other issues with any of the HOPs.

**Comment by yoyolicoris** [comment]
Thx @bohnstingl , that workaround solves my issue!

**Comment by bohnstingl** [comment]
Input mutation support for map (https://github.com/pytorch/pytorch/pull/187568) and scan (https://github.com/pytorch/pytorch/pull/186474) has landed in main.

**Comment by yoyolicoris** [comment]
Hi, it would be great to have #162906 landed as the current autograd backpropagation in `associative_scan` materialises the whole Jacobian, which is expensive.
I wrote a simple EMA filter using `associative_scan`, and for long sequences, memory consumption can easily exceed tens or even hundreds of GiB, whereas #162906 doesn't evaluate the whole matrix and keeps memory growth proportional to the sequence length. 
In my project, I ended up replicating what #162906 did to get it working, and the memory usage shrank to a few dozen MiB. It would be great to have this feature landed upstream. Thanks!

**Comment by Om-singhaI** [comment]
Hi @ydwu4 @zou3519, I would like to pick up the DTensor item in section 2, starting with `cond`, if nobody is on it.

Before writing code I want to check my reading of the current state. In main, `torch/distributed/tensor/_ops/` contains no higher order op rules, and `test_dtensor_cond` in `test/distributed/tensor/test_dtensor.py` is still marked `unittest.expectedFailure`, so `torch.cond` on DTensor still hits the assert that #164389 documented.

I also went through #164548 and the review on it, in particular the suggestion to drop the "check that both branches produce matching output placements" approach in favour of unwrapping the DTensor inputs to local tensors, having wrapper branch functions rebuild DTensors from the recorded input spec, calling `cond` on the locals, and rewrapping the result with the output spec. That is close to what `torch/_higher_order_ops/local_map.py` already does, so my instinct is to follow that precedent rather than add a strategy table.

My plan would be to stage it: `cond` in eager only, with tests and the `expectedFailure` removed, then the compile and export path, then `while_loop`, then `scan`, and only then look at factoring the common part into something reusable. I would keep the first PR small deliberately, so the deleted xfail is the proof point before anything bigger.

Three questions before I start, since each changes the design:

1. On the case raised in the review, where both branches return the same placement but the local computation is only valid under a particular input sharding, so each rank ends up with a different local value tagged `Replicate`: should that be an error, or should we redistribute to a common spec before entering the branches? This decides whether the implementation needs real sharding propagation through the subgraph or only spec bookkeeping.
2. Is this better as a `cond` specific rule, or worth writing once as a generic tensor subclass rule that `while_loop`, `scan` and the new `switch` operator could share? The review floated the generic version, and if that is the preference I would rather settle it before the first PR than refactor into it later.
3. Does `speculate_subgraph` currently trace subgraphs with DTensor inputs, or does the unwrap need to happen before Dynamo sees the subgraph? The review mentioned `local_map` hitting problems with subclass inputs there, and I could not tell from the code whether that limitation is still present.

One more, on placement: #164548 put this under `torch/distributed/tensor/_ops/`, but that directory is organised around `OpStrategy` tables for aten ops and HOPs are not aten ops. Would you rather it live next to the other `cond` dispatch entries in `torch/_higher_order_ops/cond.py`?

I am happy to scope the first PR to inference only if that is easier to review, the way the `while_loop` input mutation work was staged in #175972. I plan to develop and test on CPU with the gloo backend through `DTensorTestBase`, so I can cover replicated, `Shard(0)` and mixed placement cases without an accelerator.

If this is being done internally, or the lane turned out to be harder than it looked when it stalled last year, I would rather know now than duplicate effort. In that case I would be glad to take the efficient backward for `map` instead, porting the `HopGraphMinCutPartitioner` pattern that landed for `scan`.

**Comment by yoyolicoris** [comment]
I want to report an issue I spotted with `associative_scan`: **gradients could be wrong with multiple leaves.**
This issue already existed before #192761 landed.

How to reproduce:

```python
import torch
from torch._higher_order_ops import associative_scan

print("torch:", torch.__version__)

def affine_compose(left, right):
    """Compose two affine maps y -> a*y + b. `left` applied first:
    right(left(y)) = a_r*(a_l*y + b_l) + b_r = (a_l*a_r)*y + (a_r*b_l + b_r)"""
    a_l, b_l = left
    a_r, b_r = right
    return a_l * a_r, b_r + a_r * b_l

def naive(a, b):
    """Reference: y[t] = a[t]*y[t-1] + b[t], with y[-1] = 0."""
    y, out = torch.zeros_like(b[0]), []
    for t in range(b.shape[0]):
        y = a[t] * y + b[t]
        out.append(y)
    return torch.stack(out, 0)

torch.manual_seed(2)
a = torch.rand(5, dtype=torch.double, requires_grad=True)
b = torch.randn(5, dtype=torch.double, requires_grad=True)
gy = torch.randn_like(b)

# Scan the affine composition, keeping only the accumulated addend (= y).
# The accumulated multiplier (first leaf) is discarded, as any user of this
# scan to evaluate a linear recurrence would do.
_, y = associative_scan(affine_compose, (a, b), dim=0, combine_mode="pointwise")
ga, gb = torch.autograd.grad(y, [a, b], gy)

ar = a.detach().clone().requires_grad_(True)
br = b.detach().clone().requires_grad_(True)
ga_ref, gb_ref = torch.autograd.grad(naive(ar, br), [ar, br], gy)

print(f"forward matches loop : {torch.allclose(y.detach(), naive(a, b).detach())}")
print(f"grad_b max abs err   : {(gb - gb_ref).abs().max().item():.3e}")
print(f"grad_a max abs err   : {(ga - ga_ref).abs().max().item():.3e}")
print(f"grad_a               : {ga.tolist()}")
print(f"grad_a (reference)   : {ga_ref.tolist()}")
```

```
forward matches loop : True
grad_b max abs err   : 1.110e-16
grad_a max abs err   : 2.110e+00
grad_a               : [0.0, 0.0, -0.0, -0.0, 0.0]        <-- silently all zero
grad_a (reference)   : [0.0, 0.8884000223753735, 0.22158965684253265, 0.20621056000794483, -2.109677520659933]
```

The main cause is in this line:
https://github.com/pytorch/pytorch/blob/01d43ebff1d75014bfff1687a118014fa6a644b0/torch/_higher_order_ops/associative_scan.py#L806
Each leaf variables are treated independently, discarding inter-leaf actions like in the linear recurrence example. The accurate way is to have one compute_grad for all the leaves, so I want to call for a halt on #192762.

This was not caught in the test due to using all-one tensor
https://github.com/pytorch/pytorch/blob/01d43ebff1d75014bfff1687a118014fa6a644b0/test/functorch/test_control_flow.py#L5305-L5310
Changing to anything other than all-one will immediately break the test.

cc @bohnstingl 
The cause of this problem was found with the assistance of AI.
